crates/sui-framework/sources/governance/validator_set.move

```diff
@@ -580,6 +607,58 @@ module sui::validator_set {
         }
     }
 
+    /// Some validators' rewards may get slashed due to getting sub-par scores
+    /// for tallying rule. This function computes, using the report record, the
+    /// stake adjustment amounts of non-performant validators. When computing reward
+    /// distribution, the adjusted stake amount will be used so these validators
+    /// receive less rewards than what's proportional to their original stake.
+    fun compute_stake_adjustments(
+        self: &ValidatorSet,
+        slashed_validators: vector<address>,
+        reward_slashing_rate: u64,
+    ): (u64, VecMap<address, u64>) {
+        let total_adjustment = 0;
+        let individual_adjustments = vec_map::empty();
+        while (!vector::is_empty(&mut slashed_validators)) {
+            let validator_address = vector::pop_back(&mut slashed_validators);
+            let original_stake = validator_total_stake_amount(self, validator_address);
+            let adjustment_u128 = 
+                (original_stake as u128) * (reward_slashing_rate as u128) 
+                / BASIS_POINT_DENOMINATOR;
+            let adjustment = (adjustment_u128 as u64);
+            vec_map::insert(&mut individual_adjustments, validator_address, adjustment);
+            total_adjustment = total_adjustment + adjustment;
+        };
+        (total_adjustment, individual_adjustments)
+    }
+
+    /// Empties the validator report records of the epoch and returns the addresses of the
+    /// non-performant validators according to the input threshold. 
+    fun process_and_empty_validator_report_records(
+        self: &ValidatorSet,
+        validator_report_records: &mut VecMap<address, VecSet<address>>,
+        reward_slashing_threshold_bps: u64,
+    ): (vector<address>, u64) {
+        let num_validators = vector::length(&self.active_validators);
+        // `num_validators` can't be greater than 400 so no overflow can happen below.
+        let reward_slashing_threshold = (num_validators * reward_slashing_threshold_bps) / (BASIS_POINT_DENOMINATOR as u64);
```

Comment:
Looks like you are using validator count for the computation. Shouldn't we use their stakes, when stakes are non-uniform?

Reply:
Good point. It should be stake weighted. Will change.